README.md
# pxe-cookbook

PXE Configuration and Kickstart for Centos 6/7 servers.
This cookbook is largely based on pxe_dust by Matt Ray

## Supported Platforms

RHEL 6/7
CentOS 6/7

## Attributes

## Usage

### pxe::default

Include `pxe` in your node's `run_list`:

```json
{
  "run_list": [
    "recipe[pxe::default]"
  ]
}
```

## Contributing

1. Fork the repository on Github
2. Create a named feature branch (i.e. `add-new-recipe`)
3. Write you change
4. Write tests for your change (if applicable)
5. Run the tests, ensuring they all pass
6. Submit a Pull Request

## License and Authors

Author:: UAF-GINA (<support+chef@gina.alaska.edu>)


CHANGELOG.md
# 0.1.0

Initial release of kickstart

* Enhancements
  * an enhancement

* Bug Fixes
  * a bug fix


spec/spec_helper.rb
require 'chefspec'
require 'chefspec/deprecations'
require 'chefspec/berkshelf'

# ChefSpec::Coverage.start!

RSpec.configure do |config|
  config.formatter = :documentation
  config.color = true
  # Specify the path for Chef Solo to find cookbooks
  # config.cookbook_path = '/var/cookbooks'

  # Specify the path for Chef Solo to find roles
  # config.role_path = '/var/roles'

  # Specify the Chef log_level (default: :warn)
  # config.log_level = :debug

  # Specify the path to a local JSON file with Ohai data
  # config.path = 'ohai.json'

  # Specify the operating platform to mock Ohai data from
  # config.platform = 'ubuntu'

  # Specify the operating version to mock Ohai data from
  # config.version = '12.04'
end
# at_exit { ChefSpec::Coverage.report! }


spec/default_spec.rb
require 'spec_helper'

describe 'pxe::default' do
  let(:chef_run) do
    ChefSpec::Runner.new.converge(described_recipe)
  end


  # it 'disables the default site' do
  #
  # end

  it 'installs apache2' do
    expect(chef_run).to include_recipe('apache2')
  end

end


recipes/installers.rb
# Original Author:: Matt Ray <matt@opscode.com>
# Original Cookbook:: pxe_dust
# Author:: Scott Macfarlane
# Cookbook Name:: pxe
# Recipe:: installers
#
# Copyright 2012-2013 Opscode, Inc
#
# Licensed under the Apache License, Version 2.0 (the "License");
# you may not use this file except in compliance with the License.
# You may obtain a copy of the License at
#
#     http://www.apache.org/licenses/LICENSE-2.0
#
# Unless required by applicable law or agreed to in writing, software
# distributed under the License is distributed on an "AS IS" BASIS,
# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
# See the License for the specific language governing permissions and
# limitations under the License.
#

require 'net/http'

#location of the full stack installers
directory "#{node['pxe']['dir']}/opscode-full-stack" do
  mode '0755'
end

begin
  pxebag = data_bag('pxe')
  defaults = data_bag_item('pxe', 'default').merge(node['pxe']['defaults'])
rescue
  log "No 'pxe' data bag found" do
    level :warn
  end
end

pxebag.each do |id|
  image = defaults.merge(data_bag_item('pxe', id)).merge(node['pxe']['defaults'])

  platform = 'el'#image['platform']
  arch = image['arch']
  version = image['version'].to_i
  version = version > 6 ? 6 : version


  if image['bootstrap']
    http_proxy = image['bootstrap']['http_proxy']
    http_proxy_user = image['bootstrap']['http_proxy_user']
    http_proxy_pass = image['bootstrap']['http_proxy_pass']
    https_proxy = image['bootstrap']['https_proxy']
  end

  # only get the full stack installers to use
  rel_arch = case arch
             when 'ppc' then 'powerpc'
             when 'i386' then 'i686'
             else 'x86_64'
             end
  case version
  when /^6/
    release = "centos-6.0-#{rel_arch}"
  when /^7/
    release = "centos-6.0-#{rel_arch}"
  end

  directory "#{node['pxe']['dir']}/opscode-full-stack/#{release}" do
    mode 0755
  end

  installer = ''
  location = ''

  #for getting latest version of full stack installers
  Net::HTTP.start('www.getchef.com') do |http|
    Chef::Log.info("/chef/download?v=#{node['pxe']['chefversion']}&p=#{platform}&pv=#{version}&m=#{rel_arch}")
    response = http.get("/chef/download?v=#{node['pxe']['chefversion']}&p=#{platform}&pv=#{version}&m=#{rel_arch}")
    Chef::Log.info("Code = #{response.code}")
    location = response['location']
    Chef::Log.info("Omnitruck URL: #{location}")
    installer = location.split('/').last
    Chef::Log.debug("Omnitruck installer: #{installer}")
  end

  #download the full stack installer
  remote_file "#{node['pxe']['dir']}/opscode-full-stack/#{release}/#{installer}" do
    source location
    mode 0644
    action :create_if_missing
  end

  run_list = (image['run_list'] || '').split(',') #for supporting multiple items

  #Chef bootstrap script run by new installs
  template "#{node['pxe']['dir']}/#{id}-chef-bootstrap" do
    source 'chef-bootstrap.sh.erb'
    mode 0644
    variables(
      :release => release,
      :installer => installer,
      :interface => image['interface'] || 'eth0',
      :environment => image['environment'],
      :run_list => run_list
      )
  end

end

#link the validation_key where it can be downloaded
link "#{node['pxe']['dir']}/validation.pem" do
  to Chef::Config[:validation_key]
end


recipes/bootstrap_template.rb
# Author:: Matt Ray <matt@opscode.com>
# Cookbook Name:: pxe_dust
# Recipe:: bootstrap_template
#
# Copyright 2012-2013, Opscode, Inc
#
# Licensed under the Apache License, Version 2.0 (the "License");
# you may not use this file except in compliance with the License.
# You may obtain a copy of the License at
#
#     http://www.apache.org/licenses/LICENSE-2.0
#
# Unless required by applicable law or agreed to in writing, software
# distributed under the License is distributed on an "AS IS" BASIS,
# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
# See the License for the specific language governing permissions and
# limitations under the License.
#

require 'mixlib/shellout'

include_recipe 'pxe::common'
include_recipe 'pxe::installers'

#write out a new pxedust.erb template
remote_file "#{node['pxe']['dir']}/chef-full.erb" do
  source "https://raw.github.com/opscode/chef/master/lib/chef/knife/bootstrap/chef-full.erb"
end

# change URL from
# url="http://www.opscode.com/chef/download?v=${version}&p=${platform}&pv=${platform_version}&m=${machine}"
# to
# url="http://hypnotoad/${platform}-${platform_version}-${machine}/${filename}"
ruby_block "template url" do
  block do
    sed = "sed 's/https:\\/\\/www.opscode.com\\/chef\\//"
    sed += "http:\\/\\/#{node['ipaddress']}\\/#{node['hostname']}-/'"
    sed += " #{node['pxe']['dir']}/chef-full.erb"
    Chef::Log.debug sed
    cmd = Mixlib::ShellOut.new(sed)
    output = cmd.run_command
    Chef::Log.debug output.stdout
    nodetemplate = File.new("#{node['pxe']['dir']}/#{node['hostname']}.erb", "w+")
    nodetemplate.puts output.stdout
    nodetemplate.chmod(0644)
  end
  action :nothing
  subscribes :create, resources("remote_file[#{node['pxe']['dir']}/chef-full.erb]")
end

#write out a new install.sh
remote_file "#{node['pxe']['dir']}/original-install.sh" do
  source "https://opscode.com/chef/install.sh"
end

# change URL from
# url="http://www.opscode.com/chef/download?v=${version}&p=${platform}&pv=${platform_version}&m=${machine}"
# to
# url="http://hypnotoad/opscode-full-stack/${platform}-${platform_version}-${machine}/${filename}"
ruby_block "install.sh url" do
  block do
    sed = "sed 's/https:\\/\\/opscode.com\\/chef\\/download?v=${version}&prerelease=${prerelease}&p=${platform}&pv=${platform_version}&m=${machine}/"
    sed += "http:\\/\\/#{node['ipaddress']}\\/opscode-full-stack\\/${platform}-${platform_version}-${machine}\\/${filename}/'"
    sed += " #{node['pxe']['dir']}/original-install.sh"
    Chef::Log.debug sed
    cmd = Mixlib::ShellOut.new(sed)
    output = cmd.run_command
    Chef::Log.debug output.stdout
    nodeinstall = File.new("#{node['pxe']['dir']}/#{node['hostname']}-install.sh", "w+")
    nodeinstall.puts output.stdout
    nodeinstall.chmod(0644)
  end
  action :nothing
  subscribes :create, resources("remote_file[#{node['pxe']['dir']}/original-install.sh]")
end

#straighten up symlinks for downloading from install.sh
ruby_block "create symlinks that match up with the urls" do
  block do
    Dir.glob("#{node['pxe']['dir']}/opscode-full-stack/*").each do |distro|
      if File.directory?(distro) && !File.symlink?(distro)
        packages = Dir.glob("#{distro}/chef_*")
        versions = []
        packages.each {|x| versions << x.split('/').last.split('-')[0].split('_')[1]}
        versions.uniq.each do |version|
          vpackages = Dir.glob("#{distro}/chef_#{version}-*").sort
          vpackages.each do |filename|
            if !File.symlink?(filename)
              lname = "#{distro}/chef_#{version}_#{filename.split('_').last}"
              ln = "ln -sf #{filename} #{lname}"
              Chef::Log.debug ln
              cmd = Mixlib::ShellOut.new(ln)
              output = cmd.run_command
              Chef::Log.debug output.stdout
            end
          end
        end
      end
    end
  end
end


recipes/server.rb
include_recipe "tftp::server"
include_recipe "pxe::common"
package 'syslinux'


begin
  pxebag = data_bag('pxe')
  defaults = data_bag_item('pxe', 'default').merge(node['pxe']['defaults'])
rescue
  log "No 'pxe' data bag found" do
    level :warn
  end
end

pxebag.each do |id|
  image_dir = "#{node['tftp']['directory']}/#{id}"
  image = defaults.merge(data_bag_item('pxe', id)).merge(node['pxe']['defaults'])

  platform = image['platform']
  arch = image['arch']
  version = image['version']


  if image['root']
    root_crypted_password = image['root']['crypted_password']
  end

  directory image_dir do
    mode 0755
  end

  remote_file "#{node['pxe']['dir']}/isos/#{platform}-#{version}-#{arch}-NetInstall.iso" do
    source image['netboot_url']
    action :create_if_missing
  end

  directory "#{node['tftp']['directory']}/pxe-#{platform}-#{version}-#{arch}.0" do
    action :create
    recursive true
  end

  mount "#{node['tftp']['directory']}/pxe-#{platform}-#{version}-#{arch}.0" do
    device "#{node['pxe']['dir']}/isos/#{platform}-#{version}-#{arch}-NetInstall.iso"
    options "loop"
    action [:enable, :mount]
  end

  ruby_block 'copy pxelinux.0' do
    block do
      ::FileUtils.cp("/usr/share/syslinux/pxelinux.0", "#{node['tftp']['directory']}/#{id}/pxelinux.0")
    end
    not_if { ::File.exists?("#{node['tftp']['directory']}/#{id}/pxelinux.0") }
  end

  if image['addresses']
    image['addresses'].keys.each do |mac_address|
      mac = mac_address.gsub(/:/, '-')
      mac.downcase!

      template "#{node['tftp']['directory']}/pxelinux.cfg/01-#{mac}" do
        source 'pxelinux.cfg.erb'
        mode 0644
        variables(
          images: image
        )
      end


    end
  end
  template "#{node['pxe']['dir']}/kickstart/#{image['id']}-kickstart.cfg" do
    source "#{platform}-#{version}-kickstart.conf.erb"
    mode 0644
    variables(
      :id => id,
      :boot_volume_size => image['boot_volume_size'] || '1GB',
      :packages => image['packages'] || '',
      :root_crypted_password => image['root_crypted_password'],
      # :halt => image['halt'] || false,
      :bootstrap => image['chef'] || true,
      :root_disks => Array(image['root_disks']),
      :packages => Array(image['packages'])
      )
  end
end

link "#{node['tftp']['directory']}/pxelinux.0" do
  to 'default/pxelinux.0'
end

images = search(:pxe, '*:*' )
images = images.each do |image|
  image['platform'] ||= 'centos'
  image['version'] ||= '7.0'
  image['arch'] ||= 'x86_64'
  image['hostname'] ||= 'unknown'
  image['kickstart'] ||= "#{image['id']}-kickstart.cfg"
end


template "#{node['tftp']['directory']}/pxelinux.cfg/default" do
  source 'pxelinux.cfg.erb'
  variables(
    images: images
  )
end

include_recipe 'pxe::installers'
include_recipe 'pxe::bootstrap_template'
